File: python/sounds.py
"""Soft, warm UI cues — low fundamentals, gentle attack, long exponential
decay. Sounds like a muffled woodblock or felt-hammer tap, much more
mellow than the click-y Windows system sounds.

The start cue is a slightly higher pitch than the stop cue, so the two
events are still distinguishable without being sharp.
"""
from __future__ import annotations
import os
import tempfile
import wave
from pathlib import Path
import winsound

import numpy as np

_SR = 44_100
_DUR = 0.32  # long enough for the body of the tone to ring out


def _soft_tone(freq: float, duration: float = _DUR) -> np.ndarray:
    n = int(_SR * duration)
    t = np.arange(n, dtype=np.float32) / _SR

    # Body: fundamental + one octave above at -12 dB (warmer than 3rd-harmonic mix).
    body = (np.sin(2 * np.pi * freq * t)
            + 0.25 * np.sin(2 * np.pi * freq * 2 * t)).astype(np.float32)

    # Soft attack (~25 ms) and exponential decay.
    attack_n = int(_SR * 0.025)
    env = np.ones_like(body)
    env[:attack_n] = 0.5 - 0.5 * np.cos(np.linspace(0, np.pi, attack_n))
    decay = np.exp(-t / (duration * 0.32)).astype(np.float32)

    # 80 Hz sub thump for warmth in the first 60 ms.
    thump_n = int(_SR * 0.06)
    thump = (np.sin(2 * np.pi * 80 * t[:thump_n])
             * np.linspace(1.0, 0.0, thump_n) * 0.18).astype(np.float32)
    body[:thump_n] += thump

    return (body * env * decay * 0.18).astype(np.float32)


_CACHE = Path(tempfile.gettempdir()) / "openwritr-cues"
_CACHE.mkdir(parents=True, exist_ok=True)
_START_PATH = _CACHE / "start.wav"
_STOP_PATH = _CACHE / "stop.wav"


def _write_wav(path: Path, samples: np.ndarray) -> None:
    pcm = np.clip(samples * 32767.0, -32768, 32767).astype(np.int16)
    with wave.open(str(path), "wb") as w:
        w.setnchannels(1)
        w.setsampwidth(2)
        w.setframerate(_SR)
        w.writeframes(pcm.tobytes())


# Warm low fundamentals.
#   start ≈ G3  (196 Hz)  — friendly upward feel
#   stop  ≈ E3  (165 Hz)  — settled, conclusive
_write_wav(_START_PATH, _soft_tone(196.0))
_write_wav(_STOP_PATH,  _soft_tone(164.81))

_FLAGS = winsound.SND_FILENAME | winsound.SND_ASYNC | winsound.SND_NODEFAULT


def play_start() -> None:
    try:
        winsound.PlaySound(str(_START_PATH), _FLAGS)
    except Exception:
        pass


def play_stop() -> None:
    try:
        winsound.PlaySound(str(_STOP_PATH), _FLAGS)
    except Exception:
        pass

File: src/sounds.rs
//! Soft start/stop UI cues — generated on first use as PCM WAVs in %TEMP%,
//! played via winsound through windows API.

use parking_lot::Mutex;
use std::fs::File;
use std::io::Write;
use std::path::PathBuf;

use windows::core::PCSTR;
use windows::Win32::Media::Audio::{PlaySoundA, SND_ASYNC, SND_FILENAME, SND_NODEFAULT, SND_NOSTOP};

const SR: u32 = 44_100;

static START_PATH: Mutex<Option<PathBuf>> = Mutex::new(None);
static STOP_PATH: Mutex<Option<PathBuf>> = Mutex::new(None);

fn temp_dir() -> PathBuf { std::env::temp_dir().join("openwritr-cues") }

fn render_tone(freq: f32, duration_s: f32) -> Vec<i16> {
    let n = (SR as f32 * duration_s) as usize;
    let attack_n = (SR as f32 * 0.025) as usize;
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        let t = i as f32 / SR as f32;
        let base = (2.0 * std::f32::consts::PI * freq * t).sin();
        let octave = 0.25 * (2.0 * std::f32::consts::PI * freq * 2.0 * t).sin();
        let mut body = base + octave;
        // 60 ms sub-thump.
        if i < (SR as f32 * 0.06) as usize {
            let env_t = 1.0 - (i as f32 / (SR as f32 * 0.06));
            body += 0.18 * env_t * (2.0 * std::f32::consts::PI * 80.0 * t).sin();
        }
        let attack = if i < attack_n {
            0.5 - 0.5 * (std::f32::consts::PI * i as f32 / attack_n as f32).cos()
        } else { 1.0 };
        let decay = (-t / (duration_s * 0.32)).exp();
        let s = body * attack * decay * 0.18;
        out.push((s.clamp(-1.0, 1.0) * 32767.0) as i16);
    }
    out
}

fn write_wav(path: &PathBuf, samples: &[i16]) -> std::io::Result<()> {
    let mut f = File::create(path)?;
    let n = samples.len();
    let data_size = (n * 2) as u32;
    let total = 36 + data_size;
    f.write_all(b"RIFF")?;
    f.write_all(&total.to_le_bytes())?;
    f.write_all(b"WAVE")?;
    f.write_all(b"fmt ")?;
    f.write_all(&16u32.to_le_bytes())?;     // chunk size
    f.write_all(&1u16.to_le_bytes())?;      // PCM
    f.write_all(&1u16.to_le_bytes())?;      // mono
    f.write_all(&SR.to_le_bytes())?;
    f.write_all(&(SR * 2).to_le_bytes())?;  // byte rate
    f.write_all(&2u16.to_le_bytes())?;      // block align
    f.write_all(&16u16.to_le_bytes())?;     // bits per sample
    f.write_all(b"data")?;
    f.write_all(&data_size.to_le_bytes())?;
    for s in samples {
        f.write_all(&s.to_le_bytes())?;
    }
    Ok(())
}

fn ensure(path_slot: &Mutex<Option<PathBuf>>, name: &str, freq: f32) -> Option<PathBuf> {
    let mut slot = path_slot.lock();
    if let Some(p) = slot.as_ref() {
        if p.exists() { return Some(p.clone()); }
    }
    let dir = temp_dir();
    std::fs::create_dir_all(&dir).ok()?;
    let path = dir.join(name);
    let samples = render_tone(freq, 0.32);
    if write_wav(&path, &samples).is_ok() {
        *slot = Some(path.clone());
        Some(path)
    } else {
        None
    }
}

pub fn play_start() {
    if let Some(p) = ensure(&START_PATH, "start.wav", 196.0) {
        play(&p);
    }
}

pub fn play_stop() {
    if let Some(p) = ensure(&STOP_PATH, "stop.wav", 164.81) {
        play(&p);
    }
}

fn play(path: &PathBuf) {
    let path_str = path.to_string_lossy();
    let mut c = path_str.as_bytes().to_vec();
    c.push(0); // null-terminate
    unsafe {
        let _ = PlaySoundA(
            PCSTR(c.as_ptr()),
            None,
            SND_FILENAME | SND_ASYNC | SND_NODEFAULT | SND_NOSTOP,
        );
    }
}
